module/os/linux/spl/spl-kmem-cache.c

```diff
@@ -1185,7 +1185,6 @@ __spl_cache_grow(spl_kmem_cache_t *skc, int flags)
 		smp_mb__before_atomic();
 		clear_bit(KMC_BIT_DEADLOCKED, &skc->skc_flags);
 		smp_mb__after_atomic();
-		wake_up_all(&skc->skc_waitq);
```

Comment:
Is there any potential race associated with doing the wakeup after we've released the spinlock? Otherwise this change seems good to me.

Reply:
I would think not, because we (obviously) don't hold the spin lock while waiting, it doesn't seem like we could use it to synchronize.  I suspect that there may be other more subtle race conditions inherent to wait_event/wake_up (e.g. we're about to go to sleep, but the wake_up() already happened), but it's a bit hard to follow the twisted web of macros and assembly.  Maybe @behlendorf can enlighten us.

Reply:
Waking the waiters outside the spin lock is fine.  The `wake_up_all` was located there so that we only woke waiters when successfully allocating a new slab.  While it's unlikely this will fail it's definitely possible.  How about further updating `spl_cache_grow_work` like this to preserve this wake up behavior.

```diff
diff --git a/module/os/linux/spl/spl-kmem-cache.c b/module/os/linux/spl/spl-kmem-cache.c
index 7dd8e85..23477b0 100644
--- a/module/os/linux/spl/spl-kmem-cache.c
+++ b/module/os/linux/spl/spl-kmem-cache.c
@@ -1198,13 +1198,16 @@ spl_cache_grow_work(void *data)
        spl_kmem_alloc_t *ska = (spl_kmem_alloc_t *)data;
        spl_kmem_cache_t *skc = ska->ska_cache;
 
-       (void) __spl_cache_grow(skc, ska->ska_flags);
+       int error = __spl_cache_grow(skc, ska->ska_flags);
 
        atomic_dec(&skc->skc_ref);
        smp_mb__before_atomic();
        clear_bit(KMC_BIT_GROWING, &skc->skc_flags);
        smp_mb__after_atomic();
 
+       if (error == 0)
+               wake_up_all(&skc->skc_waitq);
+
        kfree(ska);
 }
```

Reply:
sure, done.